lib/spack/spack/build_environment.py

```diff
@@ -125,12 +127,22 @@ def __init__(self, name, jobs):
         self.jobs = jobs
 
     def __call__(self, *args, **kwargs):
+        """parallel, and jobs_env from kwargs are swallowed and used here;
+        remaining arguments are passed through to the superclass.
+        """
+
         disable = env_flag(SPACK_NO_PARALLEL_MAKE)
-        parallel = not disable and kwargs.get('parallel', self.jobs > 1)
+        parallel = (not disable) and kwargs.pop('parallel', self.jobs > 1)
 
         if parallel:
-            jobs = "-j%d" % self.jobs
-            args = (jobs,) + args
+            args = ('-j{0}'.format(self.jobs),) + args
+            jobs_env = kwargs.pop('jobs_env', None)
+            if jobs_env:
+                # Caller wants us to set an environment variable to
+                # control the parallelism.
+                make_env = kwargs.pop('env', []).copy()
+                make_env[jobs_env] = str(self.jobs)
+                kwargs['env'] = make_env
```

Comment:
In `Executable.__call__`, if `env` is set at all (including if it's an empty container), then `Executable` will use it. Overall the effect for `MakeExecutable` is that if the user:

* sets `job_env`
* does not set `env`

that the `Executable.__call__` will run with an empty environment, which changes the semantics (normally if you don't set `env`, it uses the current environment). Was that what you were going for?

Reply:
@scheibelp Genuinely not sure what to do here. In `executable.py`:
```
   env_arg = kwargs.get('env', None)
   if env_arg is None:
       env = os.environ.copy()
       env.update(self.default_env)
   else:
       env = self.default_env.copy()
       env.update(env_arg)
```

In other words, if the environment is not specified, we overlay our defaults on the current environment; otherwise, we overlay our specified environment on the defaults. It would seem that the desirable behavior in the case of specifying `jobs_env` would be, "use the environment you were always going to use, but with this added. In the case where `env` was not specified this would have to be, "The current environment with `default_env` overlaid on top and then `jobs_env` added to it." This would not, of course, be robust against changes to `executable.py`. Thoughts?

Reply:
> It would seem that the desirable behavior in the case of specifying jobs_env would be, "use the environment you were always going to use, but with this added["]

IMO that is the most sensible behavior (e.g. to keep `CC` etc.)

> but with this added. In the case where env was not specified this would have to be, "The current environment with default_env overlaid on top and then jobs_env added to it."

I think I see what you mean: to get the semantics of `default_env`, you would need to give it priority in `MakeExecutable` when `env` was not specified, if `Executable` were to change that later, `MakeExecutable` may get left behind.

I don't think you have to deal with that here. However, if desired, you could make it more robust by adding an `extra_environment` argument to `Executable.__call__` that doesn't define the whole environment but rather overlays the specified arguments on top. In that case `MakeExecutable` would be simplified.